### PR TITLE
Bump Clap 3 -> 4

### PR DESCRIPTION
**Pull request by amitdahan**
Based on #823, and adjusted breaking-changes.
Mostly drop `possible_values` in favor of using `enum`s deriving `ValueEnum`.

### CODE REVIEW

File: src/commands/env.rs [hunk 1]
```diff
@@ -163,14 +164,12 @@ mod tests {
...
-            Box::from(shell::Bash)
-        };
         Env {
-            shell: Some(shell),
-            ..Env::default()
+            #[cfg(windows)]
+            shell: Some(Shells::Cmd),
+            #[cfg(not(windows))]
+            shell: Some(Shells::Bash),
+            ..Default::default()
```
Comment (Schniz) [inline src/commands/env.rs:172]: <picture data-single-emoji=":noice:" title=":noice:">:noice:</picture>

File: src/log_level.rs [hunk 2]
```diff
@@ -37,19 +52,6 @@ impl From<LogLevel> for &'static str {
     }
 }
 
-impl std::str::FromStr for LogLevel {
-    type Err = &'static str;
-
-    fn from_str(s: &str) -> Result<LogLevel, Self::Err> {
-        match s {
-            "quiet" => Ok(Self::Quiet),
-            "info" | "all" => Ok(Self::Info),
```
Comment (Schniz) [inline src/log_level.rs:46]: I wonder if we can coerce "all" into "info" without breaking this

Comment (Schniz) [reply src/log_level.rs:46]: https://docs.rs/clap/latest/clap/_derive/index.html#possible-value-attributes

maybe this?

```rust
enum LogLevel {
  #[value(alias("all"))]
  Info
}
```

File: src/shell/shell.rs [hunk 3]
```diff
@@ -11,23 +13,38 @@ pub trait Shell: Debug {
...
 
-    fn from_str(s: &str) -> Result<Box<dyn Shell>, Self::Err> {
-        match s {
-            "cmd" => Ok(Box::from(super::windows_cmd::WindowsCmd)),
-            "zsh" => Ok(Box::from(super::zsh::Zsh)),
-            "bash" => Ok(Box::from(super::bash::Bash)),
-            "fish" => Ok(Box::from(super::fish::Fish)),
-            "powershell" => Ok(Box::from(super::powershell::PowerShell)),
-            shell_type => Err(format!("I don't know the shell type of {shell_type:?}",)),
+impl From<Shells> for Box<dyn Shell> {
```
Comment (Schniz) [inline src/shell/shell.rs:39]: This is a first good step, in a later PR we can remove `Box<dyn Shell>` entirely by implementing `Shell` for `Shells` and forward the calls

File: src/version_file_strategy.rs [hunk 4]
```diff
@@ -1,35 +1,29 @@
...
 
-#[derive(Debug, Default)]
+#[derive(Debug, Clone, Default, ValueEnum)]
 pub enum VersionFileStrategy {
+    /// Use the local version of Node defined within the current directory
     #[default]
     Local,
+    /// Use the version of Node defined within the current directory and all parent directories
     Recursive,
 }
```
Comment (Schniz) [inline src/version_file_strategy.rs:11]: <picture data-single-emoji=":keanu_thanks:" title=":keanu_thanks:">:keanu_thanks:</picture>

**Review by Schniz — COMMENTED** [review]
<picture data-single-emoji=":lgtm:" title=":lgtm:">:lgtm:</picture> overall, let's fix that `all` value and ship? <picture data-single-emoji=":lfg-intense:" title=":lfg-intense:">:lfg-intense:</picture>

File: src/shell/shell.rs (hunk 3 above)
Comment (amitdahan) [reply src/shell/shell.rs]: So kinda like this?
https://github.com/Schniz/fnm/pull/991/commits/7781b5d87354c1f9500ecfafde1aba43d6dd5e37

**Pull request merged**